libbeat/processors/add_kubernetes_metadata/cache.go

```diff
@@ -94,5 +94,18 @@ func (c *cache) cleanup() {
 }
 
 func (c *cache) stop() {
-	close(c.done)
+	if !isClosed(c.done) {
+		close(c.done)
+	}
+}
+
+// isClosed checks if a given chan is already closed
+func isClosed(ch <-chan struct{}) bool {
+	select {
+	case <-ch:
```

Comment:
@MichaelKatsoulis and @rdner I agree that we should make sure to never try to close a channel twice. 
I am not a huge fan of this `isClosed` function, even if it seems to work, it is not trivial to understand. 
Just to clarify, since `done` channel is private, nothing writes into it except for `stop(c.done)`, and only used for stopping the `cleanup` function... I would suggest to

```golang
func (c *cache) cleanup() {
	if timeout <= 0 {
		return
	}

	ticker := time.NewTicker(timeout)
	defer ticker.Stop()
	for {
		select {
		case <-c.done:
                         c.done = nil
			return
		case now := <-ticker.C:
			c.Lock()
			for k, t := range c.deleted {
				if now.After(t) {
					delete(c.deleted, k)
					delete(c.metadata, k)
				}
			}
			c.Unlock()
		}
	}
}

func (c *cache) stop() {
        if c.done != nil {
		close(c.done)
        }
}
```

Reply:
This is dangerous because if you run `cleanup` twice it will block forever on the `c.done=nil`. Since we already have a problem of running `stop` twice I think it's quite possible.

Reply:
Anyway, `k.cache = nil` is a better solution that isClosed. So I approved the PR